File: CPP/ByTier/D/1851_D_Prefix_Permutation_Sums.cpp
// Problem: CF 1851 D - Prefix Permutation Sums
// https://codeforces.com/contest/1851/problem/D

/*
 * Problem: Prefix Permutation Sums
 * 
 * Purpose: Determine if a given array of prefix sums (with one element removed)
 *          corresponds to a valid permutation of integers from 1 to n.
 * 
 * Algorithm:
 * 1. For a valid permutation, we can compute the original array elements by
 *    taking differences of consecutive prefix sums.
 * 2. Any element in the original array must be between 1 and n.
 * 3. When we subtract adjacent prefix sums to get the original elements, we may
 *    find one element which is outside the [1,n] range; that's the missing element.
 * 4. If a valid permutation exists:
 *    - There will be exactly one element > n (the missing one),
 *    - And all elements in [1,n] except the missing one must occur once.
 * 5. We use a counting array to track which numbers in [1,n] appear in the original array.
 * 6. Then, we compare against the expected values.
 * 
 * Time Complexity: O(n) per test case
 * Space Complexity: O(n) for the counting array
 */

#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;
#define ull unsigned long long
#define int long long
#define x first
#define y second
#define pii pair<int, int>
#define ar array<int, 2>
#define inf 0x3f3f3f3f
#define endl '\n'
#define _                                                                      \
  ios::sync_with_stdio(0);                                                     \
  cin.tie(0);                                                                  \
  cout.tie(0);
const int N = 2e5 + 10, mod = 1e9 + 7;
const int seed = 13331;
int n;
int a[N];
int st[N];

void solve() {
  cin >> n;
  int num1 = -1; // Stores the value of the missing element (> n)
  vector<int> v; // Stores elements that are missing from [1, n]

  // Read prefix sums
  for (int i = 1; i < n; i++)
    cin >> a[i], st[i] = 0;
  st[n] = 0;

  // Calculate original elements and find those outside [1,n]
  for (int i = 1; i < n; i++) {
    int num = a[i] - a[i - 1]; // Difference gives original element
    if (num > n)
      num1 = num; // If greater than n, it is the missing element
    else
      st[num]++; // Count occurrences of elements in [1,n]
  }

  // Find elements from [1,n] that do not appear in the original array
  for (int i = 1; i <= n; i++) {
    if (!st[i])
      v.push_back(i);
  }

  // If there is no missing element (all elements were in [1,n])
  if (num1 == -1) {
    if (v.size() == 1) {
      if (st[v[0]] == 0)
        cout << "YES" << endl;
      else
        cout << "NO" << endl;
    } else if (v.size() == 2) {
      int kk = v[0] + v[1];
      if (st[kk] == 2)
        cout << "YES" << endl;
      else
        cout << "NO" << endl;
    } else
      cout << "NO" << endl;
    return;
  }

  // If there's a missing element, check if the sum of missing elements equals it
  if (v.size() == 2) {
    if (v[0] + v[1] == num1)
      cout << "YES" << endl;
    else
      cout << "NO" << endl;
  } else
    cout << "NO" << endl;
}

signed main() {
  _ int t;
  cin >> t;
  while (t--)
    solve();
  return 0;
}


// https://github.com/VaHiX/CodeForces/

File: CSharp/ByTier/D/1851_D_Prefix_Permutation_Sums.cs
// Problem: CF 1851 D - Prefix Permutation Sums
// https://codeforces.com/contest/1851/problem/D

/*
 * Problem: Prefix Permutation Sums
 * 
 * Algorithm/Technique:
 * - The key idea is to determine whether a given array of n-1 prefix sums
 *   can be derived from a permutation of 1..n.
 * - We use a greedy approach combined with set operations to reconstruct
 *   the missing element.
 * 
 * Time Complexity: O(n) per test case, as we process each element once.
 * Space Complexity: O(n) for storing the prefix sums and the set of possible values.
 */

using System;
using System.Collections.Generic;
namespace CodeforcesConversion {
class Program {
  static void Main(string[] args) {
    int t = int.Parse(Console.ReadLine());
    while (t-- > 0) {
      int n = int.Parse(Console.ReadLine());
      List<long> arr = new List<long>();
      string[] input = Console.ReadLine().Split();
      foreach (string s in input) {
        arr.Add(long.Parse(s));
      }
      
      // HashSet to store possible values in the original permutation
      HashSet<long> st = new HashSet<long>();
      
      // Add all numbers from 1 to n to the set
      st.Add(n);
      for (int i = 0; i < n - 1; i++) {
        st.Add(i + 1);
      }
      
      int fl = 1; // flag to detect invalid cases
      long left = -1; // variable to store the missing number
      
      // Check if the first prefix sum matches a value in the set
      if (st.Contains(arr[0])) {
        st.Remove(arr[0]); // remove it from the set as it's accounted for
      } else {
        // If not, it might be the missing number
        if (left == -1)
          left = arr[0];
        else
          fl = 0; // invalid case
      }
      
      // Process the rest of the prefix sums
      for (int i = 1; i < n - 1; i++) {
        long diff = arr[i] - arr[i - 1]; // compute the difference to get a value in original array
        
        if (st.Contains(diff)) {
          st.Remove(diff); // remove the found value
        } else {
          // if not found in set, assume it is the missing number
          if (left == -1)
            left = diff;
          else {
            fl = 0; // invalid case: two missing numbers
            break;
          }
        }
      }
      
      // If any contradictions found, output NO
      if (fl == 0) {
        Console.WriteLine("NO");
      } else {
        // Compute sum of remaining numbers in set
        long sum = 0;
        foreach (var val in st)
          sum += val;
        
        // If no missing number was found earlier, compute expected sum
        if (left == -1) {
          long req = (long)n * (n + 1) / 2; // sum of first n natural numbers
          if (req - arr[n - 2] == sum) // compare with original array sum
            Console.WriteLine("YES");
          else
            Console.WriteLine("NO");
        } else {
          // If a missing number was found, check if its sum matches remaining values in the set
          if (sum == left)
            Console.WriteLine("YES");
          else
            Console.WriteLine("NO");
        }
      }
    }
  }
}
}


// https://github.com/VaHiX/CodeForces/